### PR TITLE
support assoc with custom source for Has/BelongsTo

### LINKED ISSUE #478: Allow {source, model} to be given as queryable
The idea is to allow a model to be used with different sources. We are going to support such configuration in associations too, either in the format:

    has_many :emails, {"users_emails", Email}

Or

    has_many :emails, Email, source: "users_emails"

### PR DESCRIPTION
**Pull request by tony612**
resolve #478 @josevalim I add it to `BelongsTo` association besice `Has` because I think it also needs it :)

### CODE REVIEW

**Comment by josevalim** [comment]
Yes, definitely both need it. :)

File: lib/ecto/association.ex [hunk 1]
```diff
@@ -233,11 +234,15 @@ defmodule Ecto.Association.Has do
...
     end
 
-    assoc = Keyword.fetch!(opts, :queryable)
+    queryable = Keyword.fetch!(opts, :queryable)
 
-    unless is_atom(assoc) do
-      raise ArgumentError, "association queryable must be a model, got: #{inspect assoc}"
-    end
+    assoc =
+      case queryable do
```
Comment (josevalim) [inline lib/ecto/association.ex]: What if we extract this to a function in `Ecto.Association` so both associations can use?

Comment (tony612) [inline lib/ecto/association.ex]: Yeah, I have this idea, but not sure if we need it in `BelongTo`. I'll refactor this :)

Comment (josevalim) [inline lib/ecto/association.ex]: If we don't need it in belongs to, then we don't need to worry about it. :)

**Comment by josevalim** [comment]
I have also just realized this is not complete yet. We will have to store the source from the association and, when building the association, we will need to store the source inside the model. For example, if I do this:

    build(post, :comments)

And comments is specified as:

    has_many :comments, {"my_comments", Comment}

`"my_comments"` should be stored somewhere in the built model so when we do the Repo.insert/update/delete, it will change to the proper table. The problem is that we have nowhere to store this in the model and we already have a `__state__` field, so I wanted to avoid adding more. So I thought we could do this:

1. Create a `__meta__` field that will have a `Ecto.Schema.Metadata` struct
2. The Metadata struct will have two fields: :state and :source
3. We will remove the current `__state__` to use the new `__meta__`
4. Associations and loading the resource should set the proper metadata

Since this seems to be another big task, we can do it as part of another pull request, let's get this one merged first. What do you think? Can you tackle this one too? :D

Thanks a lot, this is shaping up fantastic!

**Comment by tony612** [comment]
@josevalim Thanks for your detailed reply. I'll look into this and try it in another PR later :)

**Comment by tony612** [comment]
@josevalim Done with refactoring. I think this can be merged. Then I'll begin with the `build` thing :sunglasses:

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**